ingestion-beam/src/main/java/com/mozilla/telemetry/heka/HekaReader.java

```diff
@@ -0,0 +1,145 @@
+    Object field = message.getField(fieldDescriptor);
+    if (field instanceof List) {
+      for (Object i : (List) field) {
+        Heka.Field f = (Heka.Field) i;
+        String key = f.getName();
+        List<String> path = key.contains(".") ? Arrays.asList(key.split("\\."))
+            : Arrays.asList("meta", key);
+        String lastKey = path.get(path.size() - 1);
+        ObjectNode target = payload;
+        for (int j = 0; j < path.size() - 1; j++) {
+          String p = path.get(j);
+          target = (ObjectNode) (target.has(p) ? target.path(p) : target.putObject(p));
+        }
+        if (f.getValueType() == Heka.Field.ValueType.STRING && target.isObject()) {
+          String value = f.getValueString(0);
+          if (value.charAt(0) == '{') {
+            target.set(lastKey, Json.readObjectNode(value.getBytes(StandardCharsets.UTF_8)));
+          } else if (value.charAt(0) == '[') {
+            target.set(lastKey, Json.readArrayNode(value.getBytes(StandardCharsets.UTF_8)));
+          } else {
+            target.put(lastKey, value);
+          }
+        } else if (f.getValueType() == Heka.Field.ValueType.BOOL) {
+          target.put(lastKey, f.getValueBool(0));
+        } else if (f.getValueType() == ValueType.INTEGER) {
+          target.put(lastKey, f.getValueInteger(0));
+        } else if (f.getValueType() == Heka.Field.ValueType.DOUBLE) {
+          target.put(lastKey, f.getValueDouble(0));
+        }
+        // FIXME: do we need to handle byte fields?
+        // see code in moztelemetry and
+        // https://bugzilla.mozilla.org/show_bug.cgi?id=1339421
+      }
+    }
+
+    Map<String, String> attributes = new HashMap<>();
+    attributes.put(Attribute.DOCUMENT_NAMESPACE, "telemetry");
+    attributes.put(Attribute.HOST, message.getHostname());
+    Optional.ofNullable(payload.remove("meta")).ifPresent(meta -> {
+      Optional.ofNullable(meta.path("Date").textValue())
+          .ifPresent(s -> attributes.put(Attribute.DATE, s));
+      Optional.ofNullable(meta.path("docType").textValue())
+          .ifPresent(s -> attributes.put(Attribute.DOCUMENT_TYPE, s));
+      Optional.ofNullable(meta.path("appBuildId").textValue())
+          .ifPresent(s -> attributes.put(Attribute.APP_BUILD_ID, s));
+      Optional.ofNullable(meta.path("appName").textValue())
+          .ifPresent(s -> attributes.put(Attribute.APP_NAME, s));
+      Optional.ofNullable(meta.path("appUpdateChannel").textValue())
+          .ifPresent(s -> attributes.put(Attribute.APP_UPDATE_CHANNEL, s));
+      Optional.ofNullable(meta.path("appVersion").textValue())
+          .ifPresent(s -> attributes.put(Attribute.APP_VERSION, s));
+      Optional.ofNullable(meta.path("documentId").textValue())
+          .ifPresent(s -> attributes.put(Attribute.DOCUMENT_ID, s));
+      Optional.ofNullable(meta.path("sourceVersion").textValue())
+          .ifPresent(s -> attributes.put(Attribute.DOCUMENT_VERSION, s));
```

Comment:
@mreid-moz Is sourceVersion the doc version?

Reply:
[Yes](https://github.com/mozilla-services/lua_sandbox_extensions/blob/master/moz_telemetry/io_modules/decoders/moz_ingest/telemetry.lua#L223)